File: AlgorithmLcr/rlc_id/selector.py
"""Selector: candidate merge, equivalence clustering and ranking (DESIGN.md §5.5, §7).

Equivalence rule (decision D6):
  Two candidates (from Engine A or B) are judged equivalent if on an expanded
  validation grid (band expanded by 10x at each end, 200 log-spaced points)
      max_k |Z1(jw_k) - Z2(jw_k)| / |Z1(jw_k)| < 1e-3.

Clustering and secondary criteria:
  Equivalent candidates are merged into an equivalence class represented by
  the lowest-AICc member. Within an equivalence class, candidates are ordered
  by secondary criteria:
    1. fewer elements first,
    2. physical plausibility (elements closer to typical nominal center).
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .circuits import evaluate_f, leaf_kinds
from .fit_engine_a import Candidate

EQUIV_BAND_EXPAND = 10.0
EQUIV_N_POINTS = 200
EQUIV_MAX_REL_TOL = 1e-3


@dataclass
class EquivalenceClass:
    """A cluster of electrically equivalent candidate circuits."""

    representative: Candidate
    members: list[Candidate] = field(default_factory=list)

    @property
    def aicc(self) -> float:
        return self.representative.aicc_val

    @property
    def wrmse(self) -> float:
        return self.representative.wrmse

    @property
    def max_rel_err(self) -> float:
        return self.representative.max_rel_err

    @property
    def n_params(self) -> int:
        return self.representative.n_params


def make_validation_grid(f: np.ndarray,
                         expand: float = EQUIV_BAND_EXPAND,
                         n_points: int = EQUIV_N_POINTS) -> np.ndarray:
    """Log-spaced validation grid expanded by `expand` beyond the measured band."""
    f_min = float(np.min(f)) / expand
    f_max = float(np.max(f)) * expand
    return np.geomspace(f_min, f_max, n_points)


def are_equivalent(c1: Candidate, c2: Candidate, f_grid: np.ndarray,
                   tol: float = EQUIV_MAX_REL_TOL) -> bool:
    """Check if c1 and c2 yield indistinguishable responses on f_grid."""
    if c1.skipped or c2.skipped:
        return False
    z1 = evaluate_f(c1.tree, c1.theta, f_grid)
    z2 = evaluate_f(c2.tree, c2.theta, f_grid)
    denom = np.maximum(np.abs(z1), 1e-300)
    max_rel = float(np.max(np.abs(z1 - z2) / denom))
    return max_rel < tol


def secondary_sort_key(cand: Candidate) -> tuple[int, float]:
    """Secondary criterion: fewer elements first, then parameter plausibility."""
    # center deviation penalty: sum of (log10(v) - mid)^2
    # Typical nominal centers: R=1k (3.0), L=1mH (-3.0), C=10nF (-8.0)
    center_map = {"R": 3.0, "L": -3.0, "C": -8.0}
    penalty = sum((cand.theta[i] - center_map.get(k, 0.0)) ** 2
                  for i, k in enumerate(leaf_kinds(cand.tree)))
    return (cand.n_params, penalty)


def rank_and_cluster_equivalent(candidates: list[Candidate],
                                f: np.ndarray,
                                equiv_tol: float = EQUIV_MAX_REL_TOL,
                                n_obs: int | None = None,
                                ) -> list[EquivalenceClass]:
    """Merge equivalent candidates into equivalence classes, ranked by AICc.

    Parsimony pass (discrepancy principle): models whose per-dof RSS is
    statistically indistinguishable from the best achievable (within chi^2
    fluctuation) are all "noise-consistent"; among them the fewest-parameter
    model is promoted.  This prevents higher-order models from winning AICc
    merely by fitting noise when the true lower-order model is already at the
    noise floor.
    """
    valid = [c for c in candidates if not c.skipped and np.isfinite(c.aicc_val)]
    if not valid:
        return []

    if n_obs is None:
        n_obs = 2 * 60  # fallback; callers should pass the real 2M

    # noise-floor estimate from the best achievable per-dof RSS
    rss_per_dof = [(c.rss / max(n_obs - c.n_params, 1), c) for c in valid]
    sigma2_hat = min(r for r, _ in rss_per_dof)
    # chi^2 3-sigma upper fluctuation of the RSS/dof ratio
    margin = 3.0 * np.sqrt(2.0 / max(n_obs, 1))
    consistent = [c for r, c in rss_per_dof
                  if r <= sigma2_hat * (1.0 + margin)]
    min_p = min(c.n_params for c in consistent)
    champions = [c for c in consistent if c.n_params == min_p]
    champion = min(champions, key=lambda c: c.aicc_val)

    # Sort candidates: champion first, then remaining by AICc
    rest = [c for c in valid if c is not champion]
    rest.sort(key=lambda c: c.aicc_val)
    ordered = [champion] + rest

    # Noise-aware equivalence tolerance: with default relative weights the
    # per-point residual sigma equals sigma_rel_hat = sqrt(RSS/n_obs), and
    # two noise-floor fits of the same physical circuit can differ by a few
    # sigma at the worst grid point.  T2 equivalence (same rational function)
    # must not be split merely because of independent noise realizations.
    sigma_rel_hat = float(np.sqrt(sigma2_hat))
    eff_equiv_tol = max(equiv_tol, 3.0 * sigma_rel_hat)

    f_grid = make_validation_grid(f)

    classes: list[EquivalenceClass] = []
    for cand in ordered:
        matched = False
        for eq in classes:
            if are_equivalent(cand, eq.representative, f_grid, tol=eff_equiv_tol):
                eq.members.append(cand)
                matched = True
                break
        if not matched:
            classes.append(EquivalenceClass(representative=cand, members=[cand]))

    # Within each class, sort members by secondary criteria
    for eq in classes:
        eq.members.sort(key=secondary_sort_key)

    return classes

File: AlgorithmLcr/Try1-Completely unknown single port fitting/cppversion/src/selector.cpp
#include "selector.hpp"

#include <algorithm>
#include <cmath>
#include <map>
#include <set>

namespace rlc {

std::vector<double> makeValidationGrid(const std::vector<double>& f, double expand,
                                       int nPoints) {
    double fMin = *std::min_element(f.begin(), f.end()) / expand;
    double fMax = *std::max_element(f.begin(), f.end()) * expand;
    return geomspace(fMin, fMax, nPoints);
}

bool areEquivalent(const Candidate& c1, const Candidate& c2,
                   const std::vector<double>& fGrid, double tol) {
    if (c1.skipped || c2.skipped) return false;
    const size_t m = fGrid.size();
    std::vector<Complex> z1(m), z2(m);
    evalThetaFreq(c1.tree, c1.theta, fGrid.data(), m, z1.data());
    evalThetaFreq(c2.tree, c2.theta, fGrid.data(), m, z2.data());
    double maxRel = 0.0;
    for (size_t k = 0; k < m; ++k) {
        double rel = std::abs(z1[k] - z2[k]) / std::max(std::abs(z1[k]), 1e-300);
        maxRel = std::max(maxRel, rel);
    }
    return maxRel < tol;
}

std::pair<int, double> secondarySortKey(const Candidate& cand) {
    // center deviation penalty: sum of (log10(v) - mid)^2
    // typical nominal centers: R = 1k (3), L = 1 mH (-3), Rd = 1 ohm (0),
    // C = 10 nF (-8); iterated over PARAMETER kinds (two per L device)
    std::map<char, double> center{{'R', 3.0}, {'L', -3.0}, {'D', 0.0}, {'C', -8.0}};
    auto kinds = paramKinds(cand.tree);
    double penalty = 0.0;
    for (size_t i = 0; i < kinds.size(); ++i) {
        double c = 0.0;
        auto it = center.find(kinds[i]);
        if (it != center.end()) c = it->second;
        double dd = cand.theta[i] - c;
        penalty += dd * dd;
    }
    return {cand.nParams(), penalty};
}

std::vector<EquivalenceClass> rankAndClusterEquivalent(std::vector<Candidate> candidates,
                                                       const std::vector<double>& f,
                                                       double equivTol, int nObsIn) {
    std::vector<Candidate> valid;
    for (auto& c : candidates) {
        if (!c.skipped && std::isfinite(c.aiccVal)) valid.push_back(std::move(c));
    }
    if (valid.empty()) return {};

    int nObs = nObsIn > 0 ? nObsIn : 2 * 60;  // fallback; callers pass 2M

    // noise-floor estimate from the best achievable per-dof RSS
    double sigma2Hat = std::numeric_limits<double>::infinity();
    for (const auto& c : valid) {
        double r = c.rss / std::max((double)(nObs - c.nParams()), 1.0);
        sigma2Hat = std::min(sigma2Hat, r);
    }
    double margin = 3.0 * std::sqrt(2.0 / std::max(nObs, 1));
    std::vector<size_t> consistentIdx;
    for (size_t i = 0; i < valid.size(); ++i) {
        double r = valid[i].rss / std::max((double)(nObs - valid[i].nParams()), 1.0);
        if (r <= sigma2Hat * (1.0 + margin)) consistentIdx.push_back(i);
    }
    int minP = std::numeric_limits<int>::max();
    for (size_t i : consistentIdx) minP = std::min(minP, valid[i].nParams());
    size_t championIdx = 0;
    bool hasChampion = false;
    double bestA = std::numeric_limits<double>::infinity();
    for (size_t i : consistentIdx) {
        if (valid[i].nParams() == minP && valid[i].aiccVal < bestA) {
            bestA = valid[i].aiccVal;
            championIdx = i;
            hasChampion = true;
        }
    }

    // order: champion first, the rest by AICc (stable)
    std::vector<size_t> restIdx;
    for (size_t i = 0; i < valid.size(); ++i) {
        if (!hasChampion || i != championIdx) restIdx.push_back(i);
    }
    std::stable_sort(restIdx.begin(), restIdx.end(),
                     [&](size_t a, size_t b) { return valid[a].aiccVal < valid[b].aiccVal; });
    std::vector<Candidate> ordered;
    if (hasChampion) ordered.push_back(valid[championIdx]);
    for (size_t i : restIdx) ordered.push_back(valid[i]);

    // noise-aware equivalence tolerance (two independent noise-floor fits of
    // the same circuit must not split into two classes)
    double sigmaRelHat = std::sqrt(sigma2Hat);
    double effTol = std::max(equivTol, 3.0 * sigmaRelHat);

    std::vector<double> fGrid = makeValidationGrid(f);
    std::vector<EquivalenceClass> classes;
    for (auto& cand : ordered) {
        bool matched = false;
        for (auto& eq : classes) {
            if (areEquivalent(cand, eq.representative, fGrid, effTol)) {
                eq.members.push_back(std::move(cand));
                matched = true;
                break;
            }
        }
        if (!matched) {
            EquivalenceClass cls;
            cls.representative = cand;
            cls.members.push_back(std::move(cand));
            classes.push_back(std::move(cls));
        }
    }

    for (auto& eq : classes) {
        std::stable_sort(eq.members.begin(), eq.members.end(),
                         [](const Candidate& a, const Candidate& b) {
                             return secondarySortKey(a) < secondarySortKey(b);
                         });
    }
    return classes;
}

}  // namespace rlc
